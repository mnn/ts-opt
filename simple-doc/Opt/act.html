
<html>
<head>
<meta charset="utf-8">
<title>act (property) - ts-opt library documentation</title>
<link rel="stylesheet" href="../styles.css">

</head>
<body>
<h1>act (property)</h1>
<section class="tsd-panel tsd-member" data-className="Opt"><a id="act" class="tsd-anchor" data-debugInfo="null url"></a><h3 class="tsd-anchor-link"><span>act</span><a href=".." aria-label="Permalink" class="tsd-anchor-icon" data-origHref="#act" data-debugInfo="unknown url"><svg viewBox="0 0 24 24"><use href="#icon-anchor"></use></svg></a></h3><div class="tsd-signature"><span class="tsd-kind-property">act</span><span class="tsd-signature-symbol">:</span> <a href="../Root/_internal_.ActInClassFn.html" class="tsd-signature-type tsd-kind-interface" data-origHref="../interfaces/_internal_.ActInClassFn.html" data-debugInfo="main return | category: <<interfaces>>, fileName: <<_internal_.ActInClassFn>>, memberName: <<undefined>>">ActInClassFn</a><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type tsd-kind-type-parameter">T</span><span class="tsd-signature-symbol">&gt;</span><span class="tsd-signature-symbol"> = ...</span></div><div class="tsd-comment tsd-typography"><p>Similar to <a href="../Opt/chain.html" class="tsd-kind-method" data-origHref="Opt.html#chain" data-debugInfo="main return | category: <<undefined>>, fileName: <<Opt>>, memberName: <<chain>>">Opt.chain</a> (in other languages called <code>bind</code> or <code>&gt;&gt;=</code>), but supports more functions passed at once (resembles <code>do</code> notation in Haskell).
It is used to model a sequence of operations where each operation can fail (can return <a href=".." class="tsd-kind-class" data-origHref="_internal_.None.html" data-debugInfo="probably class | category: <<undefined>>, fileName: <<_internal_.None>>, memberName: <<undefined>>">None</a>).</p>
<pre><code class="language-ts"><span class="hl-6">// does addition when first argument is number</span><br><span class="hl-3">const</span><span class="hl-1"> </span><span class="hl-9">f1</span><span class="hl-1"> =</span><br><span class="hl-1">  (</span><span class="hl-5">x</span><span class="hl-1">: </span><span class="hl-7">string</span><span class="hl-1"> | </span><span class="hl-7">number</span><span class="hl-1">) </span><span class="hl-3">=&gt;</span><span class="hl-1"> (</span><span class="hl-5">y</span><span class="hl-1">: </span><span class="hl-7">number</span><span class="hl-1">) </span><span class="hl-3">=&gt;</span><span class="hl-1"> </span><span class="hl-0">opt</span><span class="hl-1">(</span><span class="hl-5">x</span><span class="hl-1">).</span><span class="hl-0">narrow</span><span class="hl-1">(</span><span class="hl-5">isNumber</span><span class="hl-1">).</span><span class="hl-0">map</span><span class="hl-1">(</span><span class="hl-5">z</span><span class="hl-1"> </span><span class="hl-3">=&gt;</span><span class="hl-1"> </span><span class="hl-5">z</span><span class="hl-1"> + </span><span class="hl-5">y</span><span class="hl-1">);</span><br><span class="hl-6">// passes only even numbers</span><br><span class="hl-3">const</span><span class="hl-1"> </span><span class="hl-0">f2</span><span class="hl-1"> = (</span><span class="hl-5">x</span><span class="hl-1">: </span><span class="hl-7">number</span><span class="hl-1">): </span><span class="hl-7">Opt</span><span class="hl-1">&lt;</span><span class="hl-7">number</span><span class="hl-1">&gt; </span><span class="hl-3">=&gt;</span><span class="hl-1"> </span><span class="hl-5">x</span><span class="hl-1"> % </span><span class="hl-8">2</span><span class="hl-1"> === </span><span class="hl-8">0</span><span class="hl-1"> ? </span><span class="hl-0">opt</span><span class="hl-1">(</span><span class="hl-5">x</span><span class="hl-1">) : </span><span class="hl-5">none</span><span class="hl-1">;</span><br><br><span class="hl-0">opt</span><span class="hl-1">(</span><span class="hl-8">0</span><span class="hl-1">).</span><span class="hl-0">act</span><span class="hl-1">( </span><span class="hl-6">// Some(0)</span><br><span class="hl-1">  </span><span class="hl-0">f1</span><span class="hl-1">(-</span><span class="hl-8">2</span><span class="hl-1">), </span><span class="hl-6">// Some(-2)</span><br><span class="hl-1">  </span><span class="hl-5">f2</span><span class="hl-1">, </span><span class="hl-6">// Some(-2)</span><br><span class="hl-1">  </span><span class="hl-5">optNegative</span><span class="hl-1">, </span><span class="hl-6">// None</span><br><span class="hl-1">); </span><span class="hl-6">// None</span><br><br><span class="hl-0">opt</span><span class="hl-1">(</span><span class="hl-8">0</span><span class="hl-1">).</span><span class="hl-0">act</span><span class="hl-1">( </span><span class="hl-6">// Some(0)</span><br><span class="hl-1">  </span><span class="hl-0">f1</span><span class="hl-1">(</span><span class="hl-8">1</span><span class="hl-1">), </span><span class="hl-6">// Some(1)</span><br><span class="hl-1">  </span><span class="hl-5">f2</span><span class="hl-1">, </span><span class="hl-6">// None</span><br><span class="hl-1">  </span><span class="hl-5">optNegative</span><span class="hl-1">, </span><span class="hl-6">// won't get called, still None</span><br><span class="hl-1">); </span><span class="hl-6">// None</span><br><br><span class="hl-0">opt</span><span class="hl-1">(</span><span class="hl-8">3</span><span class="hl-1">).</span><span class="hl-0">act</span><span class="hl-1">( </span><span class="hl-6">// Some(3)</span><br><span class="hl-1">  </span><span class="hl-0">f1</span><span class="hl-1">(</span><span class="hl-8">1</span><span class="hl-1">), </span><span class="hl-6">// Some(4)</span><br><span class="hl-1">  </span><span class="hl-5">f2</span><span class="hl-1">, </span><span class="hl-6">// Some(4)</span><br><span class="hl-1">  </span><span class="hl-5">optNegative</span><span class="hl-1">, </span><span class="hl-6">// Some(4)</span><br><span class="hl-1">); </span><span class="hl-6">// Some(4)</span>
</code></pre>
</div><div class="tsd-comment tsd-typography"><h4>See</h4><p><a href="../Root/act.html" class="tsd-kind-function" data-origHref="../functions/act.html" data-debugInfo="main return | category: <<functions>>, fileName: <<act>>, memberName: <<undefined>>">act</a></p>
<h4>Param</h4></div><aside class="tsd-sources"><ul><li>Defined in <a href=".." data-origHref="https://github.com/mnn/ts-opt/blob/8ad4f6f/src/Opt.ts#L355" data-debugInfo="unknown url">Opt.ts:355</a></li></ul></aside></section>
<hr/>
<a href="..">Back to index</a>
</body>
</html>
