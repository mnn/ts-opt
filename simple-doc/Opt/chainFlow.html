
<html>
<head>
<meta charset="utf-8">
<title>chainFlow (property) - ts-opt library documentation</title>
<link rel="stylesheet" href="../styles.css">

</head>
<body>
<h1>chainFlow (property)</h1>
<section class="tsd-panel tsd-member" data-className="Opt"><a id="chainFlow" class="tsd-anchor" data-debugInfo="null url"></a><h3 class="tsd-anchor-link"><span>chain<wbr>Flow</span><a href=".." aria-label="Permalink" class="tsd-anchor-icon" data-origHref="#chainFlow" data-debugInfo="unknown url"><svg viewBox="0 0 24 24"><use href="#icon-anchor"></use></svg></a></h3><div class="tsd-signature"><span class="tsd-kind-property">chain<wbr>Flow</span><span class="tsd-signature-symbol">:</span> <a href="../Root/_internal_.ActInClassFn.html" class="tsd-signature-type tsd-kind-interface" data-origHref="../interfaces/_internal_.ActInClassFn.html" data-debugInfo="main return | category: <<interfaces>>, fileName: <<_internal_.ActInClassFn>>, memberName: <<undefined>>">ActInClassFn</a><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type tsd-kind-type-parameter">T</span><span class="tsd-signature-symbol">&gt;</span><span class="tsd-signature-symbol"> = ...</span></div><div class="tsd-comment tsd-typography"><h4>Alias</h4><p><a href="../Opt/act.html" class="tsd-kind-property" data-origHref="Opt.html#act" data-debugInfo="main return | category: <<undefined>>, fileName: <<Opt>>, memberName: <<act>>">Opt.act</a></p>
<h4>See</h4><p><a href="../Root/chainFlow.html" class="tsd-kind-function" data-origHref="../functions/chainFlow.html" data-debugInfo="main return | category: <<functions>>, fileName: <<chainFlow>>, memberName: <<undefined>>">chainFlow</a></p>
<h4>Param</h4></div><aside class="tsd-sources"><ul><li>Defined in <a href=".." data-origHref="https://github.com/mnn/ts-opt/blob/8ad4f6f/src/Opt.ts#L362" data-debugInfo="unknown url">Opt.ts:362</a></li></ul></aside></section>
<hr/>
<a href="..">Back to index</a>
</body>
</html>
