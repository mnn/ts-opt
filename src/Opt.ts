const someSymbol = Symbol('Some');
const noneSymbol = Symbol('None');

/**
 * @typeparam T Wrapped value type.
 */
export abstract class Opt<T> {
  /**
   * Create Opt instance from an array of one or zero items.
   *
   * ```ts
   * Opt.fromArray([]) // None
   * Opt.fromArray([1]) // Some(1)
   * ```
   *
   * @param x
   */
  static fromArray<T>(x: [] | [T]): Opt<T> { return opt(x[0]); }

  /**
   * `false` for [[Some]], `true` for [[None]].
   */
  abstract get isEmpty(): boolean;

  /**
   * `false` for [[Some]], `true` for [[None]].
   */
  get nonEmpty(): boolean { return !this.isEmpty; }

  /**
   * Converts `Opt` to an array.
   *
   * ```ts
   * some(1).toArray() // [1]
   * none.toArray() // []
   * ```
   */
  abstract toArray(): [] | [T];

  /**
   * `1` for [[Some]], `0` for [[None]].
   */
  get length(): number { return this.isEmpty ? 0 : 1; }

  /**
   * Applies function to the wrapped value and returns a new instance of [[Some]].
   *
   * ```
   * some(1).map(x => x + 1) // Some(2)
   * none.map(x => x + 1) // None
   * ```
   *
   * @param f
   */
  abstract map<U>(f: (_: T) => U): Opt<U>;

  /**
   * Similar to [[map]], but function is expected to return [[Opt]] which will be returned.
   * Useful for including steps which may fail or return no value.
   *
   * ```
   * some(1).flatMap(x => x === 1 ? none : some(x + 1)) // None
   * none.flatMap(x => some(1)) // None
   * ```
   *
   * @param f
   */
  abstract flatMap<U>(f: (_: T) => Opt<U>): Opt<U>;

  /**
   * Alias of [[flatMap]]
   * @param f
   */
  chain<U>(f: (_: T) => Opt<U>): Opt<U> { return this.flatMap(f); }

  /**
   * Combination of [[flatMap]] and [[opt]] functions.
   *
   * ```ts
   * some(1).chainToOpt(x => x === 1 ? null : x + 1) // None
   * some(2).chainToOpt(x => x === 1 ? null : x + 1) // Some(3)
   * ```
   *
   * @param f
   */
  chainToOpt<U>(f: (_: T) => U | undefined | null): Opt<U> { return this.flatMap(x => opt(f(x))); }

  /**
   * Returns value when [[Some]], throws error with `msg` otherwise.
   * @param msg Error message.
   */
  abstract orCrash(msg: string): T;

  /**
   * Returns value for [[Some]] or `undefined` for [[None]].
   *
   * ```ts
   * some(1).orUndef() // 1
   * none.orUndef() // undefined
   * ```
   */
  abstract orUndef(): T | undefined;

  /**
   * Returns value for [[Some]] or `null` for [[None]].
   *
   * ```ts
   * some(1).orNull() // 1
   * none.orNull() // null
   * ```
   */
  abstract orNull(): T | null;

  /**
   * Returns inner value for [[Some]], `false` for [[None]].
   *
   * ```ts
   * some(1).orFalse() // 1
   * none.orFalse() // false
   * ```
   */
  abstract orFalse(): T | false;

  /**
   * Returns inner value for [[Some]], `true` for [[None]].
   *
   * ```ts
   * some(1).orTrue() // 1
   * none.orTrue() // true
   * ```
   */
  abstract orTrue(): T | true;

  /**
   * Returns inner value for [[Some]], `NaN` for [[None]].
   *
   * ```ts
   * some(1).orNaN() // 1
   * none.orNaN() // NaN
   * ```
   */
  abstract orNaN(): T | number;

  /**
   * Applies appropriate function and returns result from the function.
   *
   * ```
   * some(1).caseOf(x => x + 1, () => 0) // 2
   * none.caseOf(x => x + 1, () => 0) // 0
   * ```
   *
   * @param onSome Processing function for [[Some]].
   * @param onNone Processing function for [[None]].
   */
  abstract caseOf<R>(onSome: (x: T) => R, onNone: () => R): R;

  /**
   * Calls `f` on [[Some]] with its value, does nothing for [[None]].
   * @param f
   */
  abstract onSome(f: (x: T) => void): void;

  /**
   * Calls `f` on [[None]], does nothing for [[Some]].
   * @param f
   */
  abstract onNone(f: () => void): void;

  /**
   * Applies passed function to this instance and returns function result.
   * Also known as a function application, `|>` or pipe operator.
   *
   * ```ts
   * some(1).pipe(x => x.isEmpty) // false
   * none.pipe(x => x.isEmpty) // true
   * ```
   *
   * @param f
   */
  pipe<R>(f: (x: Opt<T>) => R): R { return f(this); }

  /**
   * Compares inner value with given value using `===`. Always `false` for [[None]].
   *
   * ```ts
   * some(1).contains(1) // true
   * some(0).contains(1) // false
   * none.contains(undefined) // false
   * ```
   *
   * @param x
   */
  abstract contains(x: T): boolean;

  /**
   * Applies `p` to inner value and passes result. Always `false` for [[None]].
   *
   * ```ts
   * some(0).exists(x => x > 0) // false
   * some(1).exists(x => x > 0) // true
   * none.exists(x => x > 0) // false
   * ```
   *
   * @param p Predicate.
   */
  abstract exists(p: (x: T) => boolean): boolean;

  /**
   * Applies `p` to inner value and passes result. Always `true` for [[None]].
   *
   * ```
   * some(0).forAll(x => x > 0) // false
   * some(1).forAll(x => x > 0) // true
   * none.forAll(x => x > 0) // true
   * ```
   * @param p Predicate.
   */
  abstract forAll(p: (x: T) => boolean): boolean;

  /**
   * Get inner value if [[Some]], or use passed `def` value for [[None]].
   *
   * ```ts
   * some(1).orElse(2) // 1
   * none.orElse(2) // 2
   * ```
   *
   * @param def Default value.
   */
  abstract orElse(def: T): T;

  /**
   * Return `this` for [[Some]], `def` for [[None]].
   *
   * ```ts
   * some(1).orElse(some(2)) // Some(1)
   * none.orElse(some(2)) // Some(2)
   * none.orElse(none) // None
   * ```
   * @param def
   */
  abstract orElseOpt(def: Opt<T>): Opt<T>;

  /**
   * Similar to [[caseOf]] but doesn't unwrap value.
   *
   * ```
   * some(1).bimap(x => x + 1, () => 0) // Some(2)
   * none.bimap(x => x + 1, () => 0) // Some(0)
   * ```
   * @param someF
   * @param noneF
   */
  abstract bimap<U>(someF: (_: T) => U, noneF: () => U): Opt<U>;

  /**
   * Similar to [[bimap]], but accepts functions returning [[Opt]].
   *
   * ```
   * some(1).flatBimap(x => some(x+1), () => some(0)) // Some(2)
   * none.flatBimap(x => some(x+1), () => some(0)) // Some(0)
   * some(5).flatBimap(x => none, () => some(0)) // None
   * ```
   * @param someF
   * @param noneF
   */
  abstract flatBimap<U>(someF: (_: T) => Opt<U>, noneF: () => Opt<U>): Opt<U>;

  /**
   * Formats [[Opt]] to string. In case of [[Some]] inner value is converted using `JSON.stringify`.
   *
   * ```ts
   * some(1).toString() // 'Some(1)'
   * none.toString() // 'None'
   * ```
   */
  abstract toString(): string;
}

class None<T> extends Opt<T> {
  readonly '@@type' = noneSymbol;

  toArray(): [] | [T] { return []; }

  get isEmpty(): boolean { return true; }

  flatMap<U>(_f: (_: T) => Opt<U>): Opt<U> { return none as unknown as Opt<U>; }

  map<U>(): Opt<U> { return none as unknown as Opt<U>; }

  orCrash(msg: string): T { throw new Error(msg); }

  orNull(): T | null { return null; }

  orUndef(): T | undefined { return undefined; }

  orFalse(): false | T { return false; }

  orTrue(): true | T { return true; }

  orNaN(): number | T { return NaN; }

  caseOf<R>(_onSome: (x: T) => R, onNone: () => R): R {
    return onNone();
  }

  onNone(f: () => void): void { f(); }

  onSome(_f: (x: T) => void): void { }

  contains(_x: T): boolean { return false; }

  exists(_p: (x: T) => boolean): boolean { return false; }

  forAll(_p: (x: T) => boolean): boolean { return true; }

  orElse(def: T): T { return def; }

  orElseOpt(def: Opt<T>): Opt<T> { return def; }

  bimap<U>(_someF: (_: T) => U, noneF: () => U): Opt<U> { return opt(noneF()); }

  flatBimap<U>(_someF: (_: T) => Opt<U>, noneF: () => Opt<U>): Opt<U> { return noneF(); }

  toString(): string { return 'None'; }
}

class Some<T> extends Opt<T> {
  readonly '@@type' = someSymbol;

  constructor(private _value: T) { super(); }

  toArray(): [] | [T] { return [this._value]; }

  get isEmpty(): boolean { return false; }

  flatMap<U>(f: (_: T) => Opt<U>): Opt<U> {
    return f(this._value);
  }

  map<U>(f: (_: T) => U): Opt<U> {
    return new Some(f(this._value));
  }

  orCrash(_msg: string): T { return this._value; }

  orNull(): T | null { return this._value; }

  orUndef(): T | undefined { return this._value; }

  orFalse(): false | T { return this._value; }

  orTrue(): true | T { return this._value; }

  orNaN(): number | T { return this._value; }

  caseOf<R>(onSome: (x: T) => R, _onNone: () => R): R { return onSome(this._value); }

  contains(x: T): boolean { return this._value === x; }

  exists(p: (x: T) => boolean): boolean { return p(this._value); }

  forAll(p: (x: T) => boolean): boolean { return p(this._value); }

  onNone(_f: () => void): void { }

  onSome(f: (x: T) => void): void { f(this._value); }

  orElse(_def: T): T { return this._value; }

  orElseOpt(_def: Opt<T>): Opt<T> { return this; }

  bimap<U>(someF: (_: T) => U, _noneF: () => U): Opt<U> { return opt(someF(this._value)); }

  flatBimap<U>(someF: (_: T) => Opt<U>, _noneF: () => Opt<U>): Opt<U> { return someF(this._value); }

  toString(): string { return `Some(${JSON.stringify(this._value)})`; }
}

const isNoneValue = (x: any): boolean => {
  return x === undefined || x === null || Number.isNaN(x);
};

/**
 * Single global instance of [[None]].
 */
export const none: None<any> = Object.freeze(new None());

/**
 * Constructs [[Some]].
 * Usually it is [[opt]] you are looking for (only in rare cases you want to have for example `Some(undefined)`).
 * @param x
 */
export const some = <T>(x: T) => Object.freeze(new Some(x));

/**
 * Main constructor function - for `undefined`, `null` and `NaN` returns [[None]].
 * Anything else is wrapped into [[Some]].
 * @param x
 */
export const opt = <T>(x: T | undefined | null): Opt<T> => isNoneValue(x) ? none : new Some(x as T);

/**
 * For falsy values returns [[None]].
 * @param x
 */
export const optFalsy = <T>(x: T | undefined | null | ''): Opt<T> => x ? new Some(x as T) : none;

/**
 * For empty array (`[]`) returns [[None]].
 * @param x
 */
export const optEmptyArray = <T>(x: T[]): Opt<T[]> => x.length ? new Some(x) : none;

/**
 * Is given value an instance of [[Opt]]?
 * @param x
 */
export const isOpt = (x: unknown): x is Opt<unknown> => x instanceof Opt;
