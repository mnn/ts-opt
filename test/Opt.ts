/* tslint:disable:no-unused-expression */

import * as chai from 'chai';
import * as spies from 'chai-spies';

import { isOpt, none, opt, Opt, optEmptyArray, optFalsy, some } from '../src/Opt';

chai.use(spies);
const {expect} = chai;
chai.should();

const add1 = (x: number) => x + 1;

const randomNumOpt = (): Opt<number> => Math.random() > .5 ? none : some(Math.random());

describe('opt', () => {
  it('construction', () => {
    expect(opt(undefined).isEmpty).to.be.true;
    expect(opt(NaN).isEmpty).to.be.true;
    expect(opt(null).isEmpty).to.be.true;
    expect(opt('').isEmpty).to.be.false;
    expect(opt('a').isEmpty).to.be.false;
    expect(opt(0).isEmpty).to.be.false;
    expect(opt(1).isEmpty).to.be.false;
    expect(opt({}).isEmpty).to.be.false;
    expect(opt([]).isEmpty).to.be.false;
  });

  it('isEmpty', () => {
    expect(opt(null).isEmpty).to.be.true;
    expect(opt(0).isEmpty).to.be.false;
  });

  it('nonEmpty', () => {
    expect(opt(null).nonEmpty).to.be.false;
    expect(opt(0).nonEmpty).to.be.true;
  });

  it('length', () => {
    expect(opt(null).length).to.eq(0);
    expect(opt(0).length).to.eq(1);
  });

  it('orUndef', () => {
    expect(opt(null).orUndef()).to.eq(undefined);
    expect(opt(0).orUndef()).to.eq(0);
  });

  it('orNull', () => {
    expect(opt(null).orNull()).to.eq(null);
    expect(opt(0).orNull()).to.eq(0);
  });

  it('orFalse', () => {
    expect(opt(null).orFalse()).to.be.false;
    expect(opt(0).orFalse()).to.eq(0);
  });

  it('orTrue', () => {
    expect(opt(null).orTrue()).to.be.true;
    expect(opt(0).orTrue()).to.eq(0);
  });

  it('orNaN', () => {
    expect(opt(null).orNaN()).to.be.NaN;
    expect(opt(0).orNaN()).to.eq(0);
  });

  it('orCrash', () => {
    expect(() => opt(null).orCrash('')).to.throw();
    expect(opt(0).orCrash('')).to.eq(0);
    expect(opt([]).orCrash('')).to.eql([]);
  });

  it('map', () => {
    expect(opt(1).map(add1).orNull()).to.eq(2);
    expect(opt(null as unknown as number).map(add1).orUndef()).to.eq(undefined);
  });

  it('flatMap', () => {
    expect(opt(1).flatMap(() => some(2)).orNull()).to.eq(2);
    expect(opt(1).flatMap(() => none).orNull()).to.eq(null);
    expect(opt(null).flatMap(() => none).orUndef()).to.eq(undefined);
    expect(opt(null).chain(() => none).orUndef()).to.eq(undefined);
  });

  it('toArray', () => {
    expect(opt(1).toArray()).to.eql([1]);
    expect(opt(null).toArray()).to.eql([]);
  });

  describe('caseOf', () => {
    it('fp', () => {
      expect(some(1).caseOf(x => 'x' + x, () => 'y')).to.eql('x1');
      expect(none.caseOf(x => 'x' + x, () => 'y')).to.eql('y');
      expect(randomNumOpt().caseOf(() => 'x', () => 'y')).to.be.oneOf(['x', 'y']);
    });
    it('imperative', () => {
      const noneCb = chai.spy();
      const someCb = chai.spy();
      noneCb.should.have.not.been.called();
      someCb.should.have.not.been.called();
      none.caseOf(someCb, noneCb);
      someCb.should.have.not.been.called();
      noneCb.should.have.been.called.once;
      some(1).caseOf(someCb, noneCb);
      someCb.should.have.been.called.once;
      someCb.should.have.been.called.always.with(1);
      noneCb.should.have.been.called.once;
    });
  });

  it('pipe', () => {
    expect(some(1).pipe(isOpt)).to.be.true;
    expect(none.pipe(isOpt)).to.be.true;
  });

  it('onNone', () => {
    const cb = chai.spy();
    cb.should.have.not.been.called();
    some(1).onNone(cb);
    cb.should.have.not.been.called();
    none.onNone(cb);
    cb.should.have.been.called.once;
  });

  it('onSome', () => {
    const cb = chai.spy();
    cb.should.have.not.been.called();
    none.onSome(cb);
    cb.should.have.not.been.called();
    some(1).onSome(cb);
    cb.should.have.been.called.once;
    cb.should.have.been.called.with.exactly(1);
  });

  it('contains', () => {
    expect(none.contains(0)).to.be.false;
    expect(some(1).contains(0)).to.be.false;
    expect(some(0).contains(0)).to.be.true;
  });

  it('exists', () => {
    expect(none.exists(() => true)).to.be.false;
    expect(none.exists(() => false)).to.be.false;
    expect(some(0).exists(() => false)).to.be.false;
    expect(some(0).exists(() => true)).to.be.true;
    const cb = chai.spy(() => true);
    none.exists(cb);
    cb.should.have.not.been.called();
    some(0).exists(cb);
    cb.should.have.been.called.with.exactly(0);
  });

  it('forAll', () => {
    expect(none.forAll(() => true)).to.be.true;
    expect(none.forAll(() => false)).to.be.true;
    expect(some(1).forAll(() => true)).to.be.true;
    expect(some(1).forAll(() => false)).to.be.false;
    const cb = chai.spy(() => true);
    none.forAll(cb);
    cb.should.have.not.been.called();
    some(0).forAll(cb);
    cb.should.have.been.called.with.exactly(0);
  });

  it('orElse', () => {
    expect(none.orElse(0)).to.be.eq(0);
    expect(some(1).orElse(0)).to.be.eq(1);
  });

  it('orElseOpt', () => {
    expect(none.orElseOpt(some(0)).orNull()).to.be.eq(0);
    expect(none.orElseOpt(none).orNull()).to.be.eq(null);
    expect(some(1).orElseOpt(none).orNull()).to.be.eq(1);
  });

  it('bimap', () => {
    expect(none.bimap(x => x, () => 'none').orNull()).to.be.eq('none');
    expect(some('some').bimap(x => x, () => 'none').orNull()).to.be.eq('some');
  });

  it('flatBimap', () => {
    expect(none.flatBimap(x => some(x), () => some('none')).orNull()).to.be.eq('none');
    expect(some('some').flatBimap(x => some(x), () => some('none')).orNull()).to.be.eq('some');
  });

  it('toString', () => {
    expect(some(1).toString()).to.eq('Some(1)');
    expect(some('a').toString()).to.eq('Some("a")');
    expect(some(null).toString()).to.eq('Some(null)');
    expect(none.toString()).to.eq('None');
  });

  it('chainToOpt', () => {
    expect(some(1).chainToOpt(x => x === 1 ? null : x + 1).orUndef()).to.be.eq(undefined);
    expect(some(2).chainToOpt(x => x === 1 ? null : x + 1).orUndef()).to.be.eq(3);
  });
});

describe('helper functions', () => {
  describe('some', () => {
    it('create', () => {
      expect(some(0).orNull()).to.eq(0);
      expect(some('').orNull()).to.eq('');
      expect(some(NaN).orNull()).to.eql(NaN);
      expect(some(null).orUndef()).to.eq(null);
    });

    it('falsy', () => {
      expect(some(null).orUndef()).to.eq(null);
      expect(some(NaN).orUndef()).to.eql(NaN);
      expect(some(undefined).orNull()).to.eq(undefined);
    });
  });

  it('fromArray', () => {
    expect(Opt.fromArray([]).orNull()).to.eq(null);
    expect(Opt.fromArray([1]).orNull()).to.eq(1);
    // expect(fromArray([1, 2]).orNull()).to.eq(1);
  });

  it('isOpt', () => {
    expect(isOpt(0)).to.be.false;
    expect(isOpt(undefined)).to.be.false;
    expect(isOpt({value: 0})).to.be.false;
    expect(isOpt(some(1))).to.be.true;
    expect(isOpt(none)).to.be.true;
  });
});

describe('optFalsy', () => {
  it('construction', () => {
    expect(optFalsy(undefined).isEmpty).to.be.true;
    expect(optFalsy(NaN).isEmpty).to.be.true;
    expect(optFalsy(null).isEmpty).to.be.true;
    expect(optFalsy('').isEmpty).to.be.true;
    expect(optFalsy('a').isEmpty).to.be.false;
    expect(optFalsy(0).isEmpty).to.be.true;
    expect(optFalsy(1).isEmpty).to.be.false;
    expect(optFalsy({}).isEmpty).to.be.false;
    expect(optFalsy([]).isEmpty).to.be.false;
  });
});

describe('optEmptyArray', () => {
  it('construction', () => {
    expect(optEmptyArray([]).isEmpty).to.be.true;
    expect(optEmptyArray([0]).isEmpty).to.be.false;
    expect(optEmptyArray([0]).orNull()).to.eql([0]);
  });
});

interface Person {
  name: string;
  surname: string | null;
}

interface Db {[_: string]: Person}

describe('examples', () => {
  it('basic', () => {
    // without
    const f = (name: string | undefined) => {
      if (!name || name === '') { throw new Error('Missing name.'); }
      return name[0];
    };

    // with
    const g = (name: string | undefined) => opt(name).orCrash('Missing name.')[0];

    f('Riker'); // 'R'
    g('Riker'); // 'R'

    // f(undefined); // exception thrown
    // g(undefined); // exception thrown

    expect(f('Riker')).to.eq('R');
    expect(g('Riker')).to.eq('R');

    expect(() => f(undefined)).to.throw();
    expect(() => g(undefined)).to.throw();
  });

  it('caseOf', () => {
    const console = {
      logHistory: [] as string[],
      log(x: string) { this.logHistory.push(x); }
    };

    // tslint:disable-next-line:no-console
    const fireMissiles = () => { console.log('FIRING!'); };
    // tslint:disable-next-line:no-console
    const printSuccess = (x: string) => { console.log(x); };

    const handleMoveVanilla = (usersMove?: string): void => usersMove ? printSuccess(usersMove) : fireMissiles();
    const handleMove = (usersMove?: string): void => opt(usersMove).caseOf(printSuccess, fireMissiles);

    handleMoveVanilla(); // prints FIRING!
    handleMove(); // prints FIRING!
    handleMoveVanilla('Build a pylon.'); // prints Build a pylon.
    handleMove('Build a pylon.'); // prints Build a pylon.

    // tslint:disable-next-line:no-console
    expect(console.logHistory).to.eql(['FIRING!', 'FIRING!', 'Build a pylon.', 'Build a pylon.']);
  });

  it('more advanced', () => {
    const db: Db = {
      0: {name: 'John', surname: null},
      1: {name: 'Worf', surname: 'Mercer'}
    };

    // without
    const f = (id: number | undefined): string | null => {
      if (id === undefined) { return null; }
      const item = db[id];
      if (!item) { return null; }
      const surname = item.surname ? item.surname.toUpperCase() : '<missing>';
      return item.name + ' ' + surname;
    };

    // with
    const g = (id: number | undefined): string | null => opt(id)
      .chainToOpt(x => db[x])
      .map(item => item.name + ' ' + opt(item.surname).map(x => x.toUpperCase()).orElse('<missing>'))
      .orNull();

    f(0); // 'John <missing>'
    g(0); // 'John <missing>'

    f(1); // 'Worf MERCER'
    g(1); // 'Worf MERCER'

    f(2); // null
    g(2); // null

    expect(f(0)).to.be.eq('John <missing>');
    expect(g(0)).to.be.eq('John <missing>');
    expect(f(1)).to.be.eq('Worf MERCER');
    expect(g(1)).to.be.eq('Worf MERCER');
    expect(f(2)).to.be.eq(null);
    expect(g(2)).to.be.eq(null);
  });
});
