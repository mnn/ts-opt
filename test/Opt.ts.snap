// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jestSnapshotSerializer renders correctly Opt prop 1`] = `
<Component
  none={None}
  null={Some(null)}
  num={Some(0)}
  obj={Some({"a":4})}
  str={Some("cat ears")}
  undef={Some(undefined)}
/>
`;
